### PR TITLE
fix(751): classify LIST leftover defaults as sockets, not pending widgets

### LINKED ISSUE #695: panel_add_node rejects every node with a required MASK input — MASK (and other core socket types) missing from SAFE_SOCKET_TYPES
### What happened / root cause
panel_add_node fails for ANY node whose required inputs include a `MASK` socket (core `InvertMask`, `SetLatentNoiseMask`, KJNodes `GrowMaskWithBlur`, `MaskToImage`, …) with:

> Required custom widget "MASK" have not registered. They may be custom widgets still loading; retry shortly.

Root cause: `web/js/lib/node-widget-materialization.js` → `SAFE_SOCKET_TYPES` (the whitelist of built-in connection datatypes that are safe to leave as sockets when no widget constructor exists) is missing `MASK`, which is a stock ComfyUI core datatype. `unavailableRequiredCustomWidgetTypes()` therefore classifies the MASK input as a possibly-still-registering custom V3 widget, `awaitRequiredCustomWidgetRegistration()` polls for 5s, and the add is rejected. Also missing from the list: `MESH`, `VOXEL`, `PHOTOMAKER` (core load3d/photomaker types) — same failure class.

Impact: the agent cannot build any masking/inpaint chain on the live canvas (every mask-consuming node is unaddable). Hit in production while adding a SetLatentNoiseMask-based product-protection refine.

### Steps to reproduce
1. Open any workflow in the panel.
2. Call `panel_add_node { class_type: "InvertMask" }` (or `SetLatentNoiseMask`, `GrowMaskWithBlur`).
3. After a ~5s poll it errors: `Required custom widget "MASK" have not registered…` — 100% reproducible; retrying never helps because MASK is a socket type, not a widget, and will never register a constructor.

### Exact error (scrubbed)
```
Error: Required custom widget "MASK" have not registered. They may be custom widgets still loading; retry shortly.
```

### Fix
applied locally: yes  ·  upstream-only: no

Add the missing core connection datatypes to `SAFE_SOCKET_TYPES`:

```diff
diff --git a/web/js/lib/node-widget-materialization.js b/web/js/lib/node-widget-materialization.js
index 1fdb801..f707a68 100644
--- a/web/js/lib/node-widget-materialization.js
+++ b/web/js/lib/node-widget-materialization.js
@@ -42,8 +42,8 @@ export function requiredWidgetInputTypes(nodeOrNodeData) {
 const SAFE_SOCKET_TYPES = new Set([
   "*", "ANY", "AUDIO", "BBOX", "CLIP", "CLIP_VISION", "CLIP_VISION_OUTPUT",
   "CONDITIONING", "CONTROL_NET", "GLIGEN", "GUIDER", "IMAGE", "IPADAPTER",
-  "LATENT", "MODEL", "NOISE", "SAMPLER", "SCHEDULER", "SIGMAS", "STYLE_MODEL",
-  "UNET", "UPSCALE_MODEL", "VAE",
+  "LATENT", "MASK", "MESH", "MODEL", "NOISE", "PHOTOMAKER", "SAMPLER",
+  "SCHEDULER", "SIGMAS", "STYLE_MODEL", "UNET", "UPSCALE_MODEL", "VAE", "VOXEL",
 ]);
```

Verified: after the patch + a frontend reload, `InvertMask`, `GrowMaskWithBlur`, and `SetLatentNoiseMask` all add cleanly and wire normally. (Note for verification: the first retry immediately after `panel_reload(scope:frontend)` can race the page navigation and still hit old module code — wait for the reload handshake before concluding the patch didn't take.)

Suggestion: consider sourcing this list from the server's known datatypes (or at least covering everything in comfy's core `IO` enum) so new core socket types don't silently break panel_add_node again.

### Environment
OS: Linux 6.17 (DGX Spark, aarch64) · ComfyUI 0.30.0 (Docker, local) · GPU: NVIDIA GB10 122GB VRAM · comfyui-mcp 0.49.8 · panel 0.11.38 (pack git 42d7f76)

### LINKED ISSUE #751: panel_add_node refuses required custom-typed SOCKETS: the allowlist in unavailableRequiredCustomWidgetTypes inverts the default
Moved from artokun/comfyui-mcp#910 — the refusal is minted here, not in the orchestrator. I traced it; the root cause is specific.

## Where it comes from

`web/js/lib/node-widget-materialization.js`:

```js
const SAFE_SOCKET_TYPES = new Set([
  "*", "ANY", "AUDIO", "BBOX", "CLIP", "CLIP_VISION", "CLIP_VISION_OUTPUT",
  "CONDITIONING", "CONTROL_NET", "GLIGEN", "GUIDER", "IMAGE", "IPADAPTER",
  "LATENT", "MODEL", "NOISE", "SAMPLER", "SCHEDULER", "SIGMAS", "STYLE_MODEL",
  "UNET", "UPSCALE_MODEL", "VAE",
]);

export function unavailableRequiredCustomWidgetTypes(nodeOrNodeData, widgetConstructors) {
  return requiredWidgetInputTypes(nodeOrNodeData).filter((type) =>
    typeof widgetConstructors?.[type] !== "function" && !SAFE_SOCKET_TYPES.has(type),
  );
}
```

A required input is refused when it has no widget constructor **and** its type is not in a hardcoded list of 23 names. That list is a closed enumeration of core connection datatypes, so:

- `VIDEO` (SaveVideo) — **core**, not on the list.
- `IMAGEUPLOAD` (LoadImage) — **core**, not on the list.
- `TRACKS_INFO` (ComfyUI-Easy-Media) — and every other pack's socket type, which the list can never contain by construction.

That matches the reporter's table exactly, including that sibling nodes from the same pack with no custom-typed required input add fine.

## The same file already has the correct rule

`missingRequiredWidgetMaterializations`, twenty lines below, documents and implements the opposite default:

```js
 * A socket-only custom datatype is deliberately not reported: it
 * has no registered widget constructor and is valid to wire later.
...
    if (!type || typeof widgetConstructors?.[type] !== "function") continue;
```

One function reads "no constructor" as *socket, fine*; the other as *unregistered widget, refuse*. Both cannot be right, and the reporter's evidence says the second one is wrong.

## Please do not just delete the allowlist

The doc comment says it guards #580's bad prompt: an unknown type is *"indistinguishable from a widget pending its extension hook"*. That concern is real — the fix needs a positive test for widget-ness rather than a negative allowlist of socket-ness. `/object_info` carries what is needed: a widget-backed input is one whose spec type is a primitive (`INT`/`FLOAT`/`STRING`/`BOOLEAN`) or a combo (type is an array of choices); anything else is a link socket. That inverts the default the safe way round and needs no list to maintain.

## One more thread worth pulling

**`IMAGEUPLOAD` is a real registered widget type**, not a socket — `LoadImage`'s upload control. If it is absent from `comfyApp.widgets`, that suggests the registry being consulted may not be where the current frontend keeps these constructors. That would independently explain why *core* nodes are affected and why the 5s `awaitRequiredCustomWidgetRegistration` poll never succeeds (the reporter retried for minutes, hard-refreshed, and re-pulled `/object_info`). Worth confirming before settling on a fix — if the registry lookup is also wrong, the allowlist is masking a second defect.

## Corroborating signal from the report

`panel_copy_nodes` + `panel_paste_nodes` of an existing instance works every time. The paste path clones an existing node instead of constructing widgets from the schema, so it never runs this check — which localises the defect to the construct path rather than to node definitions or registration timing.

Ruled out by the reporter: load race (retried over minutes), stale node defs (`panel_refresh_nodes`), stale bundle (hard refresh), missing pack.

### PR DESCRIPTION
**Pull request by artokun**
Closes #751

## Recurrence

`panel_add_node` rejected `TextEncodeQwenImageEditPlusCustom_lrzjason` because its required `configs` input is declared `LIST` (a link datatype) but also carries leftover widget metadata (`{default: null}` from Python `None`, or `{default: []}`). `inputDeclaredAsSocket` treated key *presence* as a widget declaration, so the `#626` waiver (`linkProven && socketDeclared`) never fired. The guard waited 5s for a LIST widget constructor that will never register. `panel_refresh_nodes` + retry could not help.

#1990 closed this by adding `LIST` to `SAFE_SOCKET_TYPES`. That set's type-only shortcut ignores widget-value keys entirely, so a genuine LIST widget (`{default: [1, 2]}`) would also skip the wait. Growing the allowlist is the class of fix #695 warned against.

## Change

Structural, not another name on the list:

- Leftover `default: null` / `default: []` is **not** a widget declaration. A real default (`3`, `[1, 2]`, `{}`, `""`, `false`) still is. Empty object `{}` stays a DICT widget default (#1584).
- `LIST` is a ComfyUI list-socket (`isComfyListSocketType`), the same *proof* role as `isCore3dFileType` — still held to socket-shape, not the SAFE_SOCKET_TYPES type-only shortcut.
- A type the current backend proves is an output (`knownSocketTypes`) with only leftover keys is a socket too.

`#580` unproven custom widgets, `#626` value-declaring inputs, and `#788` widget-bearing unions still wait.

## Tests

`node --test browser_tests/unit/node-widget-materialization.test.mjs` — 83 passed, including the reporter's `configs: ["LIST", {default: null}]` shape and the #580/#626/#788 intact cases.

### CODE REVIEW

**Comment by artokun** [comment]
Closing as already shipped: #751 merged as PR 1990 in v0.15.123 (`LIST` is a core socket even with leftover `default: []`; tests cover TextEncodeQwenImageEditPlusCustom_lrzjason.configs). This branch is a parallel recurrence cut against 0.15.122 and is not needed on current main.

**Comment by artokun** [comment]
Superseded by PR 1990 / v0.15.123.
